Gems/MotionMatching/Code/Source/FeatureMatrixMinMaxScaler.cpp

```diff
@@ -0,0 +1,208 @@
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Allocators.h>
+#include <fstream>
+#include <iostream>
+#include <limits>
+#include <FeatureMatrixMinMaxScaler.h>
+
+namespace EMotionFX::MotionMatching
+{
+    AZ_CLASS_ALLOCATOR_IMPL(MinMaxScaler, MotionMatchAllocator, 0)
+
+    bool MinMaxScaler::Fit(const FeatureMatrix& in, const Settings& settings)
+    {
+        const FeatureMatrix::Index numRows = in.rows();
+        const FeatureMatrix::Index numColumns = in.cols();
+
+        m_clip = settings.m_clip;
+        m_featureMin = settings.m_featureMin;
+        m_featureMax = settings.m_featureMax;
+        m_featureRange = settings.m_featureMax - settings.m_featureMin;
+        AZ_Assert(m_featureRange > s_epsilon, "Feature range too small. This will lead to divisions by infinity.");
+        m_dataMin.clear();
+        m_dataMin.resize(numColumns, std::numeric_limits<float>::max());
+        m_dataMax.clear();
+        m_dataMax.resize(numColumns, -std::numeric_limits<float>::max());
+
+        for (FeatureMatrix::Index row = 0; row < numRows; ++row)
+        {
+            for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
+            {
+                m_dataMin[column] = AZStd::min(m_dataMin[column], in(row, column));
+                m_dataMax[column] = AZStd::max(m_dataMax[column], in(row, column));
+            }
+        }
+
+        m_dataRange.clear();
+        m_dataRange.resize(numColumns);
+        for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
+        {
+            const float columnMin = m_dataMin[column];
+            const float columnMax = m_dataMax[column];
+            const float range = columnMax - columnMin;
+            m_dataRange[column] = range;
+        }
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+
+    float MinMaxScaler::Transform(float value, FeatureMatrix::Index column) const
+    {
+        const float& min = m_dataMin[column];
+        const float& range = m_dataRange[column];
+
+        float result = value;
+        if (range > s_epsilon)
+        {
+            const float normalizedValue = (value - min) / (range);
+            const float scaled = normalizedValue * m_featureRange + m_featureMin;
+
+            result = scaled;
+        }
+
+        if (m_clip)
+        {
+            result = AZ::GetClamp(result, m_featureMin, m_featureMax);
+        }
+
+        return result;
+    }
+
+    AZ::Vector2 MinMaxScaler::Transform(const AZ::Vector2& value, FeatureMatrix::Index column) const
+    {
+        return AZ::Vector2(
+            Transform(value.GetX(), column + 0),
+            Transform(value.GetY(), column + 1));
+    }
+
+    AZ::Vector3 MinMaxScaler::Transform(const AZ::Vector3& value, FeatureMatrix::Index column) const
+    {
+        return AZ::Vector3(
```

Comment:
I guess this could have a SIMD version in the future if this is a common use case and performance is an issue.

Reply:
Yeah :)